src/Silverback.Integration.Kafka/Messaging/Messages/KafkaKeyHelper.cs

```diff
@@ -1,44 +1,66 @@
-﻿// Copyright (c) 2020 Sergio Aquilini
-// This code is licensed under MIT license (see LICENSE file for details)
-
-using System;
-using System.Globalization;
-using System.Linq;
-using Silverback.Messaging.HealthChecks;
-
-namespace Silverback.Messaging.Messages
-{
-    internal static class KafkaKeyHelper
-    {
-        public static string? GetMessageKey(object? message)
-        {
-            if (message == null)
-                return null;
-
-            if (message is PingMessage)
-                return Guid.NewGuid().ToString("N", CultureInfo.InvariantCulture);
-
-            var keysDictionary =
-                message.GetType()
-                    .GetProperties()
-                    .Where(propertyInfo => propertyInfo.IsDefined(typeof(KafkaKeyMemberAttribute), true))
-                    .Select(
-                        propertyInfo => new
-                        {
-                            propertyInfo.Name,
-                            Value = propertyInfo.GetValue(message, null)?.ToString()
-                        })
-                    .ToList();
-
-            if (!keysDictionary.Any() ||
-                keysDictionary.All(keyValue => string.IsNullOrEmpty(keyValue.Value)))
-            {
-                return null;
-            }
-
-            return keysDictionary.Count == 1
-                ? keysDictionary.First().Value
-                : string.Join(",", keysDictionary.Select(p => $"{p.Name}={p.Value}"));
-        }
-    }
-}
+﻿// Copyright (c) 2020 Sergio Aquilini
+// This code is licensed under MIT license (see LICENSE file for details)
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Silverback.Messaging.HealthChecks;
+
+namespace Silverback.Messaging.Messages
+{
+    internal static class KafkaKeyHelper
+    {
+        private static readonly Dictionary<Type, PropertyInfo[]> _typeCache = new();
+
+        public static string? GetMessageKey(object? message)
+        {
+            if (message == null)
+                return null;
+
+            if (message is PingMessage)

```

Comment:
Question to you: Is this necessary?
In the `KafkaMessageKeyInitializerProducerBehavior` you create a NewGuid if there is no header.
Would save one if.